cmd/version/githubclient.go

```
@@ -0,0 +1,219 @@
+	i, _, err := c.GetTags(repo)
+	if err != nil {
+		return "", err
+	}
+
+	return i.TagName, nil
+}
+
+func NewGithubClient(apiKey string, httpClient *http.Client) (*GithubClient, error) {
+	if httpClient == nil {
+		httpClient = http.DefaultClient
+	}
+
+	url, err := url.Parse(baseURL)
+	if err != nil {
+		return nil, err
+	}
+	client := &GithubClient{
+		apiKey:  apiKey,
+		baseURL: url,
+		client:  httpClient,
+	}
+
+	return client, nil
+}
+
+type GithubClient struct {
+	client  *http.Client
+	apiKey  string
+	baseURL *url.URL
+}
+
+type ErrorResponse struct {
+	StatusCode int
+	Errors     string
+}
+
+func (c *GithubClient) DoRequest(method, path string, body, v interface{}) (*ErrorResponse, error) {
+	req, err := c.NewRequest(method, path, body)
+	if err != nil {
+		return nil, err
+	}
+
+	return c.Do(req, v)
+}
+
+func (c *GithubClient) NewRequest(method, path string, body interface{}) (*http.Request, error) {
+	// relative path to append to the endpoint url, no leading slash please
+	if path[0] == '/' {
+		path = path[1:]
+	}
+	rel, err := url.Parse(path)
+	if err != nil {
+		return nil, err
+	}
+
+	u := c.baseURL.ResolveReference(rel)
+	var req *http.Request
+	if body != nil {
+		bodyBytes, _ := json.Marshal(body)
+		req, _ = http.NewRequest(method, u.String(), bytes.NewBuffer(bodyBytes))
+		req.Header.Set("Content-Type", "application/json")
+	} else {
+		req, _ = http.NewRequest(method, u.String(), nil)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	req.Close = true
+
+	req.Header.Add("MC-Api-Key", c.apiKey)
+	return req, nil
+}
+
+func (c *GithubClient) Do(req *http.Request, v interface{}) (*ErrorResponse, error) {
+	resp, err := c.client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+
+	defer resp.Body.Close()
+	if resp.StatusCode > 299 {
+		o, _ := io.ReadAll(resp.Body)
+		errResp := &ErrorResponse{
+			StatusCode: resp.StatusCode,
+			Errors:     string(o),
+		}
+
+		return errResp, fmt.Errorf("%s returned %d", req.URL, resp.StatusCode)
```

I don't get this API... This returns an `error` but also a response...
`*ErrorResponse` should instead implement `error` and return just that. Not both values.

---

there is no need for this actually

---

all errors are ignored here anyways